sled-hardware/src/disk.rs

```diff
@@ -341,23 +404,91 @@ impl Disk {
 
     // Ensure that the zpool contains all the datasets we would like it to
     // contain.
-    fn ensure_zpool_has_datasets(
+    async fn ensure_zpool_has_datasets(
+        log: &Logger,
         zpool_name: &ZpoolName,
+        disk_identity: &DiskIdentity,
+        key_requester: Option<&StorageKeyRequester>,
     ) -> Result<(), DiskError> {
-        let datasets = match zpool_name.kind().into() {
-            DiskVariant::M2 => M2_EXPECTED_DATASETS.iter(),
-            DiskVariant::U2 => U2_EXPECTED_DATASETS.iter(),
+        let (root, datasets) = match zpool_name.kind().into() {
+            DiskVariant::M2 => (None, M2_EXPECTED_DATASETS.iter()),
+            DiskVariant::U2 => {
+                (Some(CRYPT_DATASET), U2_EXPECTED_DATASETS.iter())
+            }
         };
-        for dataset in datasets.into_iter() {
+
+        let zoned = false;
+        let do_format = true;
+
+        // Ensure the root encrypted filesystem exists
+        // Datasets below this in the hierarchy will inherit encryption
+        if let Some(dataset) = root {
+            let Some(key_requester) = key_requester else {
+                return Err(DiskError::MissingStorageKeyRequester);
+            };
             let mountpoint = zpool_name.dataset_mountpoint(dataset);
+            let keypath: Keypath = disk_identity.into();
+
+            let epoch =
+                if let Ok(epoch_str) = Zfs::get_oxide_value(dataset, "epoch") {
+                    if let Ok(epoch) = epoch_str.parse::<u64>() {
+                        epoch
+                    } else {
+                        return Err(DiskError::MissingEpochProperty(
+                            dataset.to_string(),
+                        ));
+                    }
+                } else {
+                    // We assume the error indicates that the dataset doesn't
+                    // exist here. Use the latest secret to create it. If the
+                    // error did not actually indicate that the dataset didn't
+                    // exist the creation will fail, so we don't need to do any
+                    // special handling.
+                    key_requester.load_latest_secret().await?
```

Comment:
What happens if this assumption is wrong? For example, if we have a dataset where the epoch property has been deleted?\

Reply:
Perhaps I was not really clear in this comment. I'll rewrite it. The third sentence states that if the dataset does actually exist, the attempt to create it immediately below will fail.

Reply:
For some reason I missed the second part of your sentence. If the epoch value was deleted, the call to `Zfs::get_oxide_value` will succeed. The epoch will just be represented by a `-` and won't be parsable.

I updated the comment in [580259d](https://github.com/oxidecomputer/omicron/pull/3007/commits/580259d853c2f1fa4668e040e074e6c6c8f0a0a7) to clarify.